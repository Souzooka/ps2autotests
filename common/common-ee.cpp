#include <stdio.h>
#include <stdarg.h>
#include <string.h>
#include <kernel.h>

#define TEST_THREAD_PRIORITY   0x40
#define TEST_THREAD_STACK_SIZE 0x8000

char schedfBuffer[65536];
unsigned int schedfBufferPos = 0;

void schedf(const char *format, ...) {
	va_list args;
	va_start(args, format);
	schedfBufferPos += vsprintf(schedfBuffer + schedfBufferPos, format, args);
	va_end(args);
}

void flushschedf() {
	printf("%s", schedfBuffer);
	schedfBuffer[0] = '\0';
	schedfBufferPos = 0;
}

u32 g_testThreadDoneSema = 0;

extern int test_main(int argc, char *argv[]);

char testThreadStack[TEST_THREAD_STACK_SIZE] __attribute__ ((aligned(16)));
void testThreadProc() {
	//TODO: Print begin/end traces
	//TODO: Pass arguments from main thread to the test function
	test_main(0, NULL);
	SignalSema(g_testThreadDoneSema);
}

int main(int argc, char *argv[]) {
	//A thread is created to ensure the execution environment is the same
	//across all possible boot methods (direct hw boot, ps2link, emulator, etc.)
	
	ee_sema_t semaInfo;
	memset(&semaInfo, 0, sizeof(ee_sema_t));
	semaInfo.init_count = 0;
	semaInfo.max_count  = 1;
	semaInfo.option     = 0;
	g_testThreadDoneSema = CreateSema(&semaInfo);
	
	u32 testThreadId = 0;
	ee_thread_t threadParam;
	memset(&threadParam, 0, sizeof(ee_thread_t));
	threadParam.func               = (void*)&testThreadProc;
	threadParam.initial_priority   = TEST_THREAD_PRIORITY;
	threadParam.stack_size         = TEST_THREAD_STACK_SIZE;
	threadParam.stack              = testThreadStack;
	testThreadId = CreateThread(&threadParam);
	
	StartThread(testThreadId, NULL);
	WaitSema(g_testThreadDoneSema);
	
	return 0;
}
