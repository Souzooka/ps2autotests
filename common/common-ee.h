#ifndef PS2AUTOTESTS_COMMON_EE_H
#define PS2AUTOTESTS_COMMON_EE_H

// Defines for MSVC highlighting.  Not intended to actually compile with msc.
#ifdef _MSC_VER
#define __STDC__
#define _EE
#define __attribute__(x)
#endif

#include <stdio.h>
#include <tamtypes.h>

#undef main
#define main test_main

void schedf(const char *format, ...);
void flushschedf();

#endif
