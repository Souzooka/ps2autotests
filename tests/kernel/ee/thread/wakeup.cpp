#include <common-ee.h>
#include <kernel.h>
#include <string.h>

#define STACK_SIZE 0x8000

char nullThreadStack[STACK_SIZE] __attribute__ ((aligned(16)));
void nullThreadProc(u32) {
	//Do nothing
}

char sleepingThreadStack[STACK_SIZE] __attribute__ ((aligned(16)));
void sleepingThreadProc(u32) {
	SleepThread();
}

char waitingThreadStack[STACK_SIZE] __attribute__ ((aligned(16)));
void waitingThreadProc(u32) {
	ee_sema_t dummySemaInfo;
	memset(&dummySemaInfo, 0, sizeof(ee_sema_t));
	dummySemaInfo.init_count = 0;
	dummySemaInfo.max_count  = 1;
	dummySemaInfo.option     = 0;
	u32 dummySema = CreateSema(&dummySemaInfo);
	WaitSema(dummySema);
}

int createTestThread(void *entry, int prio, void *stack, u32 stackSize) {
	ee_thread_t threadParam;
	memset(&threadParam, 0, sizeof(ee_thread_t));
	threadParam.func             = entry;
	threadParam.initial_priority = prio;
	threadParam.stack            = stack;
	threadParam.stack_size       = stackSize;
	return CreateThread(&threadParam);
}

void printThreadStatus(int threadId) {
	ee_thread_status_t threadStat;
	memset(&threadStat, 0, sizeof(ee_thread_status_t));
	int result = ReferThreadStatus(threadId, &threadStat);
	schedf("result: %02x, init prio: %02x, current prio: %02x, status: %02x, wait type: %d, wakeup count: %d\n", 
		result, threadStat.initial_priority, threadStat.current_priority, threadStat.status, threadStat.waitType, threadStat.wakeupCount);
}

void doWakeupThread(int threadId) {
	int result = WakeupThread(threadId);
	if (result >= 0) {
		schedf("succeeded -> result: %s, status -> ", (result == threadId) ? "thread id" : "something else");
		printThreadStatus(threadId);
	} else {
		schedf("failed -> result: %d\n", result);
	}
}

int main(int argc, char *argv[]) {
	printf("-- TEST BEGIN\n");

	int testThreadPriority = getThreadPriority(GetThreadId());

	{
		schedf("self thread:\n");
		
		schedf("  wakeup (with tid 0): ");
		doWakeupThread(0);
		
		schedf("  wakeup (with current tid): ");
		doWakeupThread(GetThreadId());
		
		flushschedf();
	}

	{
		schedf("worse prio thread:\n");
		
		int threadId = createTestThread((void*)&nullThreadProc, testThreadPriority + 0x10, nullThreadStack, STACK_SIZE);
		schedf("  wakeup before start: ");
		doWakeupThread(threadId);

		StartThread(threadId, NULL);
		schedf("  wakeup after start: ");
		doWakeupThread(threadId);
		
		SuspendThread(threadId);
		schedf("  wakeup after suspend: ");
		doWakeupThread(threadId);
		
		TerminateThread(threadId);
		schedf("  wakeup after terminate: ");
		doWakeupThread(threadId);
		
		DeleteThread(threadId);

		flushschedf();
	}

	{
		schedf("sleeping thread:\n");
		
		int threadId = createTestThread((void*)&sleepingThreadProc, testThreadPriority - 0x10, sleepingThreadStack, STACK_SIZE);

		StartThread(threadId, 0);
		schedf("  wakeup after start: ");
		doWakeupThread(threadId);

		//Restart thread because it was woken up
		StartThread(threadId, 0);

		SuspendThread(threadId);
		schedf("  wakeup after suspend: ");
		doWakeupThread(threadId);
		
		TerminateThread(threadId);
		DeleteThread(threadId);

		flushschedf();
	}

	{
		schedf("waiting thread:\n");

		int threadId = createTestThread((void*)&waitingThreadProc, testThreadPriority - 0x10, waitingThreadStack, STACK_SIZE);

		StartThread(threadId, 0);
		schedf("  wakeup after start: ");
		doWakeupThread(threadId);
		
		SuspendThread(threadId);
		schedf("  wakeup after suspend: ");
		doWakeupThread(threadId);
		
		TerminateThread(threadId);
		DeleteThread(threadId);

		flushschedf();
	}

	{
		schedf("corner cases:\n");
		
		schedf("  wakeup with invalid tid: ");
		doWakeupThread(~0);

		flushschedf();
	}

	printf("-- TEST END\n");
	
	return 0;
}
