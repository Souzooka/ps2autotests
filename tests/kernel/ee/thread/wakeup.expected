-- TEST BEGIN
self thread:
  wakeup (with tid 0): failed -> result: -1
  wakeup (with current tid): failed -> result: -1
worse prio thread:
  wakeup before start: failed -> result: -1
  wakeup after start: succeeded -> result: thread id, status -> result: 02, init prio: 50, current prio: 50, status: 02, wait type: 0, wakeup count: 1
  wakeup after suspend: succeeded -> result: thread id, status -> result: 08, init prio: 50, current prio: 50, status: 08, wait type: 0, wakeup count: 2
  wakeup after terminate: failed -> result: -1
sleeping thread:
  wakeup after start: succeeded -> result: thread id, status -> result: 10, init prio: 30, current prio: 30, status: 10, wait type: 0, wakeup count: 0
  wakeup after suspend: succeeded -> result: thread id, status -> result: 08, init prio: 30, current prio: 30, status: 08, wait type: 0, wakeup count: 0
waiting thread:
  wakeup after start: succeeded -> result: thread id, status -> result: 04, init prio: 30, current prio: 30, status: 04, wait type: 2, wakeup count: 1
  wakeup after suspend: succeeded -> result: thread id, status -> result: 0c, init prio: 30, current prio: 30, status: 0c, wait type: 2, wakeup count: 2
corner cases:
  wakeup with invalid tid: failed -> result: -1
-- TEST END
